app/models/asset.rb

```diff
@@ -52,6 +52,14 @@ class Asset
     event :scanned_infected do
       transition any => :infected
     end
+
+    event :upload_success do
+      transition %i(clean uploaded not_uploaded) => :uploaded
+    end
+
+    event :upload_failure do
+      transition %i(clean uploaded not_uploaded) => :not_uploaded
```

Comment:
Did you consider adding the `uploaded` state only, and using `clean` to mean that it's been virus scanned but not uploaded? Is there any benefit to knowing that the upload has been tried and failed over knowing that the upload has not yet succeeded?

Reply:
No, I hadn't considered that. I think it's a good idea.